core/trino-main/src/main/java/io/trino/operator/project/MergePages.java

```diff
@@ -75,12 +75,59 @@ private MergePages() {}
             WorkProcessor<Page> pages,
             AggregatedMemoryContext memoryContext)
     {
-        return pages.transform(new MergePagesTransformation(
-                types,
-                minPageSizeInBytes,
-                minRowCount,
-                maxPageSizeInBytes,
-                memoryContext.newLocalMemoryContext(MergePages.class.getSimpleName())));
+        List<Type> typeList = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        WorkProcessor.Transformation<Page, Page> mergingTransform;
+        if (typeList.isEmpty()) {
+            // position count only mode
+            mergingTransform = new MergePositionCountPagesTransformation(minRowCount);
+        }
+        else {
+            mergingTransform = new MergePagesTransformation(
+                    typeList,
+                    minPageSizeInBytes,
+                    minRowCount,
+                    maxPageSizeInBytes,
+                    memoryContext.newLocalMemoryContext(MergePages.class.getSimpleName()));
+        }
+        return pages.transform(mergingTransform);
+    }
+
+    private static class MergePositionCountPagesTransformation
+            implements WorkProcessor.Transformation<Page, Page>
+    {
+        private final int minRowCount;
+        private int currentRowCount;
+
+        private MergePositionCountPagesTransformation(int minRowCount)
+        {
+            this.minRowCount = minRowCount;
+        }
+
+        @Override
+        public TransformationState<Page> process(Page inputPage)
+        {
+            if (inputPage == null) {
+                if (currentRowCount == 0) {
+                    return finished();
+                }
+                Page result = new Page(currentRowCount);
+                currentRowCount = 0;
+                return ofResult(result, false);
+            }
+
+            if (currentRowCount == 0 && inputPage.getPositionCount() >= minRowCount) {
+                // reuse input page directly as the output
+                return ofResult(inputPage);
+            }
+
+            currentRowCount += inputPage.getPositionCount();
+            if (currentRowCount >= minRowCount) {
+                Page result = new Page(currentRowCount);
+                currentRowCount = 0;
+                return ofResult(result);
```

Comment:
when pages carry no data, why are we producing "as small pages as possible" instead of producing as big ones as possible?

Reply:
I had the same thought, just saw this comment and pushed a revision that targets 8k batches (minSize controls the preference to reuse the input page still)

Reply:
potentially related
https://github.com/trinodb/trino/blob/b77dd122029a42bbf80ba9c9fd8853daf4522da9/plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/CountQueryPageSource.java#L27-L30
https://github.com/trinodb/trino/pull/3512#discussion_r416862305

Reply:
Definitely conceptually related, but slightly different in which scenarios it affects. This addresses the case where the entire output projection is only a row count and emitted pages from the projection are smaller the minimum row count threshold- in that scenario, the page builder size will not change but the row count (once it does finally flush) can have arbitrarily many rows (the effect of this can end up being very similar to problem discussed in the linked PR).